File: patterns/_tests/StructuralPatterns/Facade/java/integration.test.java
// ============================================================
//  Integration Tests — Facade: Third Party API Hub
//
//  What we test here:
//    · Full makePurchase() produces correct end-to-end output
//    · Auth result feeds into Payment, Payment ref feeds into Messaging
//    · Each facade routes exclusively to its own TP base URL
//    · Credentials masked with ••• — no raw values in output
//    · Swapping a facade subclass leaves other facades untouched
//
//  Run: javac -cp .:junit-platform-console-standalone.jar integration.test.java && java ...
// ============================================================

import org.junit.jupiter.api.*;
import java.io.*;
import java.util.*;
import static org.junit.jupiter.api.Assertions.*;

public class integration {

    static List<String> capture(Runnable fn) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        PrintStream old = System.out;
        System.setOut(new PrintStream(baos));
        try { fn.run(); } finally { System.setOut(old); }
        return Arrays.asList(baos.toString().split("\\r?\\n"));
    }

    // ── Full purchase flow ────────────────────────────────────
    @Nested
    class FullPurchaseFlow {
        List<String> lines;

        @BeforeEach
        void setup() {
            lines = capture(() -> new PurchaseService().makePurchase("tok_user_jane", 49.99, "+15550142"));
        }

        @Test
        void all_three_confirmations_present() {
            String out = String.join("\n", lines);
            assertTrue(out.contains("✓ User:"));
            assertTrue(out.contains("✓ Charged"));
            assertTrue(out.contains("✓ Message queued"));
        }

        @Test
        void auth_email_in_output() {
            assertTrue(lines.stream().anyMatch(l -> l.contains("jane.doe@example.com")));
        }

        @Test
        void transaction_id_in_output() {
            assertTrue(lines.stream().anyMatch(l -> l.contains("txn_8821")));
        }

        @Test
        void message_id_in_output() {
            assertTrue(lines.stream().anyMatch(l -> l.contains("msg_3301")));
        }
    }

    // ── Loose coupling ────────────────────────────────────────
    @Nested
    class LooseCoupling {
        @Test
        void each_facade_routes_to_its_own_base_url() {
            var paymentLines = capture(() -> new PaymentFacade().charge("usr_1", 100));
            var authLines    = capture(() -> new AuthFacade().validateToken("tok_1"));
            var msgLines     = capture(() -> new MessagingFacade().send("+1", "hi"));

            assertTrue(paymentLines.stream().anyMatch(l -> l.contains("paymentprocessortp.com")));
            assertTrue(authLines.stream().anyMatch(l -> l.contains("authprovidertp.com")));
            assertTrue(msgLines.stream().anyMatch(l -> l.contains("messagingtp.com")));
        }

        @Test
        void no_cross_contamination_between_urls() {
            String paymentOut = String.join("\n", capture(() -> new PaymentFacade().charge("usr_1", 100)));
            String authOut    = String.join("\n", capture(() -> new AuthFacade().validateToken("tok_1")));

            assertFalse(paymentOut.contains("authprovidertp.com"));
            assertFalse(paymentOut.contains("messagingtp.com"));
            assertFalse(authOut.contains("paymentprocessortp.com"));
        }

        @Test
        void headers_masked_raw_credentials_never_appear() {
            String out = String.join("\n", capture(() ->
                new PurchaseService().makePurchase("tok", 10, "+1")));
            assertFalse(out.contains("sk_live_demo"));
            assertFalse(out.contains("auth_secret_demo"));
            assertFalse(out.contains("SID_xyz"));
            assertTrue(out.contains("•••"));
        }
    }

    // ── Single point of change ───────────────────────────────
    @Nested
    class SinglePointOfChange {
        @Test
        void swapping_payment_facade_does_not_affect_auth() {
            // Custom payment facade returns fixed response
            PaymentFacade freePayment = new PaymentFacade() {
                @Override
                public Map<String, Object> charge(String userId, double amount) {
                    return map("transaction_id", "free_txn", "status", "success");
                }
            };

            var authResult = new AuthFacade().validateToken("tok_1");
            assertEquals("usr_4421", authResult.get("user_id"));

            var tx = freePayment.charge("usr_1", 0);
            assertEquals("free_txn", tx.get("transaction_id"));
        }

        @Test
        void swapping_messaging_does_not_affect_payment() {
            MessagingFacade silentMsg = new MessagingFacade() {
                @Override
                public Map<String, Object> send(String to, String message) {
                    return map("message_id", "silent", "status", "dropped");
                }
            };

            var tx = new PaymentFacade().charge("usr_1", 50);
            assertEquals("txn_8821", tx.get("transaction_id"));

            var msg = silentMsg.send("+1", "test");
            assertEquals("dropped", msg.get("status"));
        }
    }
}

File: patterns/_tests/StructuralPatterns/Facade/csharp/integration.test.cs
// ============================================================
//  Integration Tests — Facade: Third Party API Hub
//
//  What we test here:
//    · Full MakePurchase() produces correct end-to-end output
//    · Auth result feeds into Payment, Payment ref feeds into Messaging
//    · Each facade routes exclusively to its own TP base URL
//    · Credentials masked with ••• — no raw values in output
//    · Swapping a facade subclass leaves other facades untouched
//
//  Run: dotnet test  (requires xUnit project referencing the source)
// ============================================================

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Xunit;

public class FacadeIntegrationTests
{
    static List<string> Capture(Action fn)
    {
        var sb  = new System.Text.StringBuilder();
        var old = Console.Out;
        Console.SetOut(new StringWriter(sb));
        try { fn(); } finally { Console.SetOut(old); }
        return sb.ToString().Split('\n', StringSplitOptions.RemoveEmptyEntries)
                            .Select(l => l.TrimEnd('\r')).ToList();
    }

    // ── Full purchase flow ────────────────────────────────────
    public class FullPurchaseFlowTests
    {
        private readonly List<string> _lines = Capture(() =>
            new PurchaseService().MakePurchase("tok_user_jane", 49.99, "+15550142"));

        [Fact]
        public void All_Three_Confirmations_Present()
        {
            string out = string.Join("\n", _lines);
            Assert.Contains("✓ User:",         out);
            Assert.Contains("✓ Charged",       out);
            Assert.Contains("✓ Message queued", out);
        }

        [Fact]
        public void Auth_Email_In_Output() =>
            Assert.Contains(_lines, l => l.Contains("jane.doe@example.com"));

        [Fact]
        public void Transaction_Id_In_Output() =>
            Assert.Contains(_lines, l => l.Contains("txn_8821"));

        [Fact]
        public void Message_Id_In_Output() =>
            Assert.Contains(_lines, l => l.Contains("msg_3301"));
    }

    // ── Loose coupling ────────────────────────────────────────
    public class LooseCouplingTests
    {
        [Fact]
        public void Each_Facade_Routes_To_Its_Own_BaseUrl()
        {
            var paymentLines = Capture(() => new PaymentFacade().Charge("usr_1", 100));
            var authLines    = Capture(() => new AuthFacade().ValidateToken("tok_1"));
            var msgLines     = Capture(() => new MessagingFacade().Send("+1", "hi"));

            Assert.Contains(paymentLines, l => l.Contains("paymentprocessortp.com"));
            Assert.Contains(authLines,    l => l.Contains("authprovidertp.com"));
            Assert.Contains(msgLines,     l => l.Contains("messagingtp.com"));
        }

        [Fact]
        public void No_Cross_Contamination_Between_URLs()
        {
            string paymentOut = string.Join("\n", Capture(() => new PaymentFacade().Charge("usr_1", 100)));
            string authOut    = string.Join("\n", Capture(() => new AuthFacade().ValidateToken("tok_1")));

            Assert.DoesNotContain("authprovidertp.com",    paymentOut);
            Assert.DoesNotContain("messagingtp.com",       paymentOut);
            Assert.DoesNotContain("paymentprocessortp.com", authOut);
        }

        [Fact]
        public void Headers_Masked_Raw_Credentials_Never_Appear()
        {
            string out = string.Join("\n", Capture(() =>
                new PurchaseService().MakePurchase("tok", 10, "+1")));
            Assert.DoesNotContain("sk_live_demo",     out);
            Assert.DoesNotContain("auth_secret_demo", out);
            Assert.DoesNotContain("SID_xyz",          out);
            Assert.Contains("•••", out);
        }
    }

    // ── Single point of change ───────────────────────────────
    public class SinglePointOfChangeTests
    {
        sealed class FreePaymentFacade : PaymentFacade
        {
            public override Dictionary<string, object> Charge(string userId, double amount) =>
                new() { ["transaction_id"] = "free_txn", ["status"] = "success" };
        }

        sealed class SilentMessagingFacade : MessagingFacade
        {
            public override Dictionary<string, object> Send(string to, string message) =>
                new() { ["message_id"] = "silent", ["status"] = "dropped" };
        }

        [Fact]
        public void Swapping_Payment_Does_Not_Affect_Auth()
        {
            var authResult = new AuthFacade().ValidateToken("tok_1");
            Assert.Equal("usr_4421", authResult["user_id"]);

            var tx = new FreePaymentFacade().Charge("usr_1", 0);
            Assert.Equal("free_txn", tx["transaction_id"]);
        }

        [Fact]
        public void Swapping_Messaging_Does_Not_Affect_Payment()
        {
            var tx = new PaymentFacade().Charge("usr_1", 50);
            Assert.Equal("txn_8821", tx["transaction_id"]);

            var msg = new SilentMessagingFacade().Send("+1", "test");
            Assert.Equal("dropped", msg["status"]);
        }
    }
}
